PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h

```diff
@@ -1119,65 +1057,79 @@ void ResetEventByEventQuantities()
-
-  // Centrality: TBI
+  //   *) SelectedTracks: => cut directly in void process( ... )
+  //   *) Centrality: TBI
   //  if ((  TBI   < ceh_a.fEventCuts[eCentrality][eMin]) || (  TBI   >
   //  ceh_a.fEventCuts[eCentrality][eMax])) {
   //    return kFALSE;
   //  }
-
-  // Vertex_x:
-  if ((collision.posX() < ceh_a.fEventCuts[eVertex_x][eMin]) ||
-      (collision.posX() > ceh_a.fEventCuts[eVertex_x][eMax])) {
+  //   *) Vertex_x:
+  if ((collision->posX() < ceh_a.fEventCuts[eVertex_x][eMin]) ||
+      (collision->posX() > ceh_a.fEventCuts[eVertex_x][eMax])) {
     return kFALSE;
   }
-
-  // Vertex_y:
-  if ((collision.posY() < ceh_a.fEventCuts[eVertex_y][eMin]) ||
-      (collision.posY() > ceh_a.fEventCuts[eVertex_y][eMax])) {
+  //   *) Vertex_y:
+  if ((collision->posY() < ceh_a.fEventCuts[eVertex_y][eMin]) ||
+      (collision->posY() > ceh_a.fEventCuts[eVertex_y][eMax])) {
     return kFALSE;
   }
-
-  // Vertex_z:
-  if ((collision.posZ() < ceh_a.fEventCuts[eVertex_z][eMin]) ||
-      (collision.posZ() > ceh_a.fEventCuts[eVertex_z][eMax])) {
+  //   *) Vertex_z:
+  if ((collision->posZ() < ceh_a.fEventCuts[eVertex_z][eMin]) ||
+      (collision->posZ() > ceh_a.fEventCuts[eVertex_z][eMax])) {
     return kFALSE;
   }
 
-  // NContributors:
-  if ((collision.numContrib() < ceh_a.fEventCuts[eNContributors][eMin]) ||
-      (collision.numContrib() > ceh_a.fEventCuts[eNContributors][eMax])) {
-    return kFALSE;
-  }
+  // b) Event cuts on info available only in reconstructed data:
+  if (gProcessRec || gProcessRecSim) {
+    CollisionRec const* collisionRec = reinterpret_cast<CollisionRec const*>(collision);
+    // CollisionRec const& collision = *collisionRec;
+    //   *) NContributors:
+    if ((collisionRec->numContrib() < ceh_a.fEventCuts[eNContributors][eMin]) ||
+        (collisionRec->numContrib() > ceh_a.fEventCuts[eNContributors][eMax])) {
+      return kFALSE;
+    }
+    // ...
+  } // if (gProcessRec || gProcessRecSim) {
+
+  // c) Event cuts on info available only in simulated data:
+  if (gProcessSim) {
+    CollisionSim const* collisionSim = reinterpret_cast<CollisionSim const*>(collision);
+    // CollisionSim const& collision = *collisionSim;
+    //   *) Impact parameter:
+    if ((collisionSim->impactParameter() < ceh_a.fEventCuts[eImpactParameter][eMin]) ||
+        (collisionSim->impactParameter() > ceh_a.fEventCuts[eImpactParameter][eMax])) {
+      return kFALSE;
+    }
+    // ...
+  } // if (gProcessSim) {
 
   return kTRUE;
 
-} // template <typename T> Bool_t EventCuts(aod::Collision const& collision, T const& tracks)
+} // template <typename T1, typename T2> Bool_t EventCuts(T1 const* collision, T2 const* tracks)
 
 //============================================================
 
 template <eRecSim rs, typename T1, typename T2>
-void FillEventHistograms(T1 const& collision, T2 const& tracks, eBeforeAfter ba)
+void FillEventHistograms(T1 const* collision, T2 const* tracks, eBeforeAfter ba)
```

Comment:
That's the great point in using templates!
You can keep the reference

Reply:
I switched to pointers, because of casting. 

In general, I am not against templates. But also, I do not want to pack my analysis code ONLY with templates, C++ offers many more functionalities that we should cherish as well...